server/src/main/java/io/spine/server/AbstractServiceBuilder.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2022, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import io.spine.reflect.GenericTypeIndex;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static io.spine.util.Exceptions.newIllegalStateException;
+
+/**
+ * Abstract for builder of service classes.
+ *
+ * @param <T>
+ *         the type of service to be produced
+ * @param <B>
+ *         the self-type of the builder for return type covariance
+ */
+public abstract class AbstractServiceBuilder<T, B extends AbstractServiceBuilder<T, B>> {
+
+    private final Set<BoundedContext> contexts = new HashSet<>();
```

Comment:
Let's make sure we have both `equals()` and `hashCode()` defined for `BoundedContext`.

From what I can see now, we don't.

Reply:
I made them comparable by the names too. We use this already at least once, for diagnostics.